caper/resource_analysis.py

```diff
@@ -0,0 +1,338 @@
+
+    def __init__(self, metadata_jsons):
+        """Solves y = f(x) in a statistical way where
+        x is a vector of input file sizes (e.g. size_each([bam, bowtie2_index_tar, ...]))
+        y is a vector of resources (e.g. [max_mem, max_disk, ...])
+
+        self._tasks is extended (across all workflows) list of resource monitoring
+        result from gcp_monitor():
+        [
+            {
+                'task_name': TASK1,
+                'workflow_id': WORKFLOW_ID_OF_TASK1,
+                ...
+            },
+            {
+                'task_name': TASK2,
+                'workflow_id': WORKFLOW_ID_OF_TASK2,
+                ...
+            }
+        ]
+        Args:
+            metadata_jsons:
+                List of metadata JSON file URIs or metadata JSONs
+                or CromwellMetadata objects.
+        """
+        self._tasks = []
+        for metadata_json in metadata_jsons:
+            if isinstance(metadata_json, CromwellMetadata):
+                cm = metadata_json
+            else:
+                cm = CromwellMetadata(metadata_json)
+            self._tasks.extend(cm.gcp_monitor())
+
+    def analyze(
+        self,
+        in_file_vars=None,
+        reduce_in_file_vars=DEFAULT_REDUCE_IN_FILE_VARS,
+        target_resources=DEFAULT_TARGET_RESOURCES,
+        plot_pdf=None,
+    ):
+        """Find and analyze all tasks.
+
+        Args:
+            in_file_vars:
+                Dict { TASK_NAME: [IN_FILE_VAR1, ...] }.
+                If not None, only tasks defined in it will be analyzed.
+                See ResourceAnalysis.analyze_per_task.__doc__ for details.
+            reduce_in_file_vars:
+                Python function (e.g. sum, max) to reduce x matrix into a vector.
+                See ResourceAnalysis.analyze_per_task.__doc__ for details.
+            target_resources:
+                Keys (in dot notation) to make vector y.
+            plot_pdf:
+                Local file name for a PDF plot.
+        Returns:
+            Results in a dict form: {
+                TASK_NAME: {
+                    'x': X_DATA,
+                    'y': Y_DATA,
+                    'coeffs': ANALYSIS_RESULT
+                }
+            }
+        """
+        plot_pp = None
+        if plot_pdf:
+            from matplotlib.backends.backend_pdf import PdfPages
+
+            plot_pp = PdfPages(plot_pdf)
+
+        result = {}
+
+        if in_file_vars:
+            all_tasks = in_file_vars.keys()
+        else:
+            all_tasks = self._get_all_task_names()
+
+        for task_name in all_tasks:
+            result[task_name] = self.analyze_task(
+                task_name,
+                in_file_vars=in_file_vars[task_name] if in_file_vars else None,
+                reduce_in_file_vars=reduce_in_file_vars,
+                target_resources=target_resources,
+                plot_pp=plot_pp,
+            )
+
+        if plot_pdf:
+            plot_pp.close()
```

Comment:
You don't really use `plot_pp` except to pass to `analyze_task`. Seems like it would make sense to manage that object's lifecycle within `analyze_task` instead of here.

Reply:
All tasks' analysis result is written on one multi-page PDF file.
Not sure if we can append pages to an existing PDF file.
https://stackoverflow.com/questions/27419001/is-it-possible-to-append-figures-to-matplotlibs-pdfpages/27419170#27419170

So I think we cannot manage lifecycle of the plot PDF object in `analyze_task`.

Reply:
gotcha, makes sense.